edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.DiagnosticsComponent/MetricsWorker.cs

```diff
@@ -0,0 +1,125 @@
+
+        async Task Scrape(CancellationToken cancellationToken)
+        {
+            using (await this.scrapeUploadLock.LockAsync(cancellationToken))
+            {
+                Log.LogInformation("Scraping Metrics");
+                List<Metric> metricsToPersist = new List<Metric>();
+                int numScrapedMetrics = 0;
+                foreach (var scrapedMetric in await this.scraper.ScrapeEndpointAsync(cancellationToken))
+                {
+                    numScrapedMetrics++;
+                    // Get the previous scrape for this metric
+                    if (this.metrics.TryGetValue(scrapedMetric.GetMetricKey(), out Metric oldMetric))
+                    {
+                        // If the metric is unchanged, do nothing
+                        if (oldMetric.Value.Equals(scrapedMetric.Value))
+                        {
+                            continue;
+                        }
+
+                        // if the metric has changed, write the previous metric to disk
+                        metricsToPersist.Add(oldMetric);
+                    }
+
+                    // if new metric or metric changed, save to local buffer
+                    this.metrics[scrapedMetric.GetMetricKey()] = scrapedMetric;
+                }
+
+                Log.LogInformation($"Scraped {numScrapedMetrics} Metrics");
+
+                if (metricsToPersist.Count != 0)
+                {
+                    Log.LogInformation("Storing Metrics");
+                    this.storage.WriteData(Newtonsoft.Json.JsonConvert.SerializeObject(metricsToPersist));
+                    Log.LogInformation("Stored Metrics");
+                }
+            }
+        }
+
+        async Task Upload(CancellationToken cancellationToken)
+        {
+            using (await this.scrapeUploadLock.LockAsync(cancellationToken))
+            {
+                Log.LogInformation("Uploading Metrics");
+                await this.uploader.UploadAsync(this.GetMetricsToUpload(this.lastUploadTime), cancellationToken);
+                Log.LogInformation("Uploaded Metrics");
+
+                this.storage.RemoveOldEntries(this.lastUploadTime);
+                this.metrics.Clear();
+                this.lastUploadTime = this.systemTime.UtcNow;
+            }
+        }
+
+        IEnumerable<Metric> GetMetricsToUpload(DateTime lastUploadTime)
+        {
+            // Get all metrics that have been stored since the last upload
+            foreach (KeyValuePair<DateTime, Func<string>> data in this.storage.GetData(lastUploadTime))
+            {
+                var fileMetrics = Newtonsoft.Json.JsonConvert.DeserializeObject<Metric[]>(data.Value()) ?? Enumerable.Empty<Metric>();
+                foreach (Metric metric in fileMetrics)
+                {
+                    yield return metric;
+                }
+            }
+
+            // Get all metrics stored in the local buffer.
+            foreach (Metric metric in this.metrics.Values)
```

Comment:
if some metrics ares stored in-memory, there will be possible that these metrics will be lost due to crash/unexpected restart.  All metric should save to persisted storage after scraping.

Reply:
This makes sense. I was originally only saving the old metric if the new one was different, but I can instead have both saved to disk and just have a copy of the most recent metric in memory. Changed to reflect this.